Adds a new `ResPath` struct that will replace `ResourcePath`

Ygg01 opened the pull request:
Wanted some feedback from maintainers. 

Stuff to do:

- [x] Benchmarks
- [x] Decide on `ResPathSpan` a wrapper struct for `ReadOnlySpan`.
- [x] Decide what to do with construction. Keep same constructor as `ResourcePath`. Decided: No.
- [x] ~~Fast creation - via factory method.~~ Pointless from talki with PJB
- [x] ~~Avoid branching in fast factory method?~~ Yes.
- [x] Docs.
- [x] Tests. 


This relates to #2540
It doesn't fix because it would require wholesale replacement. 

No longer a breaking change.

mirrorcult commented:
if this PR isnt gonna be breaking should at least obsolete `ResourcePath`

mirrorcult reviewed (approved):
looks fine to me especially with the test but need benchmarks imo

Ygg01 commented:
Added Benchmarks and fix some issues using a custom enumerator.

```
|           Method |   N |        Mean |     Error |    StdDev |
|----------------- |---- |------------:|----------:|----------:|
|      ReadResPath |   1 |    349.3 ns |   6.78 ns |   9.05 ns |
|  ReadResPathFast |   1 |    186.7 ns |   3.75 ns |   5.14 ns |
| ReadResourcePath |   1 |    470.0 ns |   9.37 ns |  15.92 ns |
|      ReadResPath |  10 |  1,864.6 ns |  36.71 ns |  64.29 ns |
|  ReadResPathFast |  10 |    265.9 ns |   5.23 ns |   8.14 ns |
| ReadResourcePath |  10 |  2,572.3 ns |  50.37 ns |  51.73 ns |
|      ReadResPath | 100 | 17,050.7 ns | 332.89 ns | 444.40 ns |
|  ReadResPathFast | 100 |  1,075.6 ns |  21.23 ns |  34.87 ns |
| ReadResourcePath | 100 | 23,655.3 ns | 351.22 ns | 328.53 ns |
```

AJCM-git commented on Robust.Shared/Utility/ResourcePath.cs:
```diff
@@ -15,6 +15,7 @@ namespace Robust.Shared.Utility
     ///     ResourcePaths are immutable.
     /// </summary>
     [PublicAPI, Serializable, NetSerializable]
+    [Obsolete]
     public sealed class ResourcePath : IEquatable<ResourcePath>
```
Obsolete the public fields/methods in the class with a message mentioning what their replacement in ResPath is

Ygg01 commented:
Removed branching from fast constructor. Added type serializer.

Speed is now: 
|                     Method |   N |           Mean |       Error |      StdDev |
|--------------------------- |---- |---------------:|------------:|------------:|
| CreateWithSeparatorResPath |   1 |    110.2079 ns |   1.2814 ns |   1.1986 ns |
|    CreateUnsafePathResPath |   1 |      0.0000 ns |   0.0000 ns |   0.0000 ns |
|         CreateResourcePath |   1 |    246.2445 ns |   4.7017 ns |   4.3980 ns |
| CreateWithSeparatorResPath |  10 |  1,041.1877 ns |   7.1897 ns |   6.7253 ns |
|    CreateUnsafePathResPath |  10 |      1.9065 ns |   0.0072 ns |   0.0068 ns |
|         CreateResourcePath |  10 |  1,296.2405 ns |   7.3435 ns |   6.5098 ns |
| CreateWithSeparatorResPath | 100 | 10,423.0894 ns |  61.4379 ns |  57.4691 ns |
|    CreateUnsafePathResPath | 100 |     24.3702 ns |   0.3385 ns |   0.3167 ns |
|         CreateResourcePath | 100 | 12,240.8429 ns | 234.7124 ns | 270.2951 ns |

However, it will allocate a bunch:
`ResourcePath`:
![H1GdXWtGB6](https://user-images.githubusercontent.com/1146204/214072342-c274ddd6-ed86-491d-b4f1-c667e5ca45de.png)
![U7RyablEqC](https://user-images.githubusercontent.com/1146204/214072349-27a148b2-3e77-48ee-937e-57ec322a338a.png)

`ResPath`:
![mjKsk83NLA](https://user-images.githubusercontent.com/1146204/214072410-6452df17-6cb3-49a6-895b-cb280900b912.png)
![ra8OPiuwlj](https://user-images.githubusercontent.com/1146204/214072417-6549173e-0dbf-4b31-acfc-c26e05d3e73e.png)

Some pooling or just using Refs could be a workable solution. This is why I made factory methods rather than constructors.

AlexMorgan3817 commented:
More of formatting. And... Pendulum is based

AlexMorgan3817 commented on Robust.Shared/Utility/ResPath.cs:
```diff
@@ -0,0 +1,657 @@
+
+    public const string SystemSeparatorStr = @"\";
+#else
+    public const char SystemSeparator = '/';
```
```suggestion
    public const char SystemSeparator = '/';

```

Ygg01 replied:
I don't see any changes here. I added a new line after block.

DrSmugleaf commented on Robust.Benchmarks/Utility/ResourcePathBench.cs:
```diff
@@ -0,0 +1,59 @@
+
+    [UsedImplicitly]
+    [Params(10, 100, 1000)]
+    public int N;
```
N is commonly used for the amount of iterations in a benchmark
https://github.com/dotnet/BenchmarkDotNet

AlexMorgan3817 replied:
> N is commonly used for the amount of iterations in a benchmark https://github.com/dotnet/BenchmarkDotNet

Yes, but what is iterations, do you just iterate integer numbers? :\

AlexMorgan3817 replied:
Okay, anyway, maybe i am just not experienced in C# enough

AlexMorgan3817 commented on Robust.Shared/Utility/ResPath.cs:
```diff
@@ -22,12 +22,12 @@ namespace Robust.Shared.Utility;
 #else
     public const char SystemSeparator = '/';
     public const string SystemSeparatorStr = "/";
 #endif
```
```suggestion
    public const char SystemSeparator = '/';

    public const string SystemSeparatorStr = "/";
#endif
```
As you wish

metalgearsloth commented on Robust.Shared/Utility/ResPath.cs:
```diff
@@ -0,0 +1,680 @@
+    {
+    }
+
+    /// <summary>
+    ///     Create a new path from a string, splitting it by the separator provided.
+    /// </summary>
+    /// <param name="path">The string path to turn into a resource path.</param>
+    /// <param name="separator">The separator for the resource path. If null or empty it will default to <c>/</c></param>
+    /// <exception cref="ArgumentException">Thrown if you try to use "." as separator.</exception>
+    public static ResPath CreateWithSeparator(string path, char separator = '/')
+    {
+        if (separator is '.')
+        {
+            throw new ArgumentException("Separator may not be `.`  Prefer \\ or /");
+        }
+
+        if (path == "" || path == ".")
+        {
+            return Self;
+        }
+
+        var sb = new StringBuilder(path.Length, path.Length * 2);
+        var segments = path.Split(separator);
+        if (path[0] == separator) sb.Append('/');
+
+        var needsSeparator = false;
+        foreach (var segment in segments)
+        {
+            if ((segment == "." && segments.Length != 0) || segment == "") continue;
+
+            if (needsSeparator) sb.Append('/');
+
+            sb.Append(segment);
+            needsSeparator = true;
+        }
+
+        return sb.Length == 0
+            ? Self
+            : new ResPath(sb.ToString());
+    }
```
Is CreateWIthSeparator an accurate name anymore?

Ygg01 replied:
It's a placeholder name for constructor taking path with arbitrary separators. E.g. `a|b|c` and separator `|` result in `/a/b/c` `CanonPath`.

Ygg01 replied:
Removed per PJB comment.

metalgearsloth commented on Robust.Shared/Utility/ResPath.cs:
```diff
@@ -0,0 +1,680 @@
+    }
+
+
+    /// <summary>
+    ///     Enumerates the segments of this path.
+    /// </summary>
+    /// <remarks>
+    ///     Segments are returned from highest to deepest.
+    ///     For example <c>/a/b</c> will yield <c>a</c> then <c>b</c>.
+    ///     No special indication is given for rooted paths,
+    ///     so <c>/a/b</c> yields the same as <c>a/b</c>.
+    /// </remarks>
+    public IEnumerable<string> EnumerateSegments()
+    {
+        foreach (var fragment in CanonPath.Split(Separator))
+        {
+            yield return fragment;
+        }
+    }
```
Should be a struct enumerator.

Ygg01 replied:
Removed per PJB's comment.

metalgearsloth commented on Robust.Shared/Utility/ResPath.cs:
```diff
@@ -0,0 +1,680 @@
+    /// <returns>if true if path wouldn't be modified </returns>
+    public bool IsClean()
+    {
+        var segments = CanonPath.Split(Separator).ToArray();
```
Is the ToArray needed?

metalgearsloth commented on Robust.Shared/Utility/ResPath.cs:
```diff
@@ -0,0 +1,680 @@
+        {
+            if (segments[i] != "..") continue;
+
+            if (IsRooted)
+            {
+                return false;
+            }
```
Is this supposed to be in the loop?

Ygg01 replied:
Honestly, it's verbatim of `ResourcePath` method, I really had no luck figuring it out at the time. That said, I maybe could try and refactor it, possibly remove the array.

metalgearsloth commented on Robust.Shared/Utility/ResPath.cs:
```diff
@@ -0,0 +1,680 @@
+    /// </remarks>
+    public ResPath Clean()
+    {
+        var segments = new List<string>();
```
ValueList, should also be constructed after the if check.

metalgearsloth reviewed (changes requested):
PJB should also take a look.

PJB3005 reviewed (changes requested):
Much of the complexity kept from `ResourcePath` isn't necessary and needs to be removed. The maintainer meeting for this was a year ago but we agreed all the complexities with "regular filesystem" semantics like `.` and `..` traversal should just be axed. Really the only thing you need at most is like, concatenation taking root vs relative into account.

metalgearsloth commented on Robust.Shared/Utility/ResPath.cs:
```diff
@@ -0,0 +1,477 @@
+    /// </summary>
+    public static readonly ResPath Self = new(".");
+
+    /// <summary>
+    ///     "/" (root) as a static. Separator used is <c>/</c>.
+    /// </summary>
+    public static readonly ResPath Root = new("/");
+
+    /// <summary>
+    ///     "/" (root) as a static. Separator used is <c>/</c>.
+    /// </summary>
+    public static readonly ResPath Empty = new("");
```
doc summary looks copied?

metalgearsloth commented on Robust.Shared/Utility/ResourcePath.cs:
```diff
@@ -15,6 +15,8 @@ namespace Robust.Shared.Utility
     ///     ResourcePaths are immutable.
     /// </summary>
     [PublicAPI, Serializable, NetSerializable]
+    [Obsolete(@$"Use {nameof(ResPath)} instead, for extra performance
+        use {nameof(ResPath)} but only if '/' is the separator")]
```
Comment looks doubled? Should just say to use ResPath instead.

Ygg01 commented on Robust.Shared/Serialization/TypeSerializers/Implementations/ResPathSerializer.cs:
```diff
@@ -0,0 +1,80 @@
+            path /= "meta.json";
+        }
+
+        if (!path.CanonPath.Split('/').First().Equals("Textures", StringComparison.InvariantCultureIgnoreCase))
+        {
+            path = SharedSpriteComponent.TextureRootResPath / path;
+        }
+
+        path = path.ToRootedPath();
```
Shrug, no idea; I copied to make sure serialization worked in-game (when I briefly replaced `ResourcePath` with `ResPath`).

https://github.com/space-wizards/RobustToolbox/blob/40c8a3ccf78bd3f9b8f9554242e3bf373c968f89/Robust.Shared/Serialization/TypeSerializers/Implementations/ResourcePathSerializer.cs#L40-L43
